detector.go

```diff
@@ -1,118 +1,161 @@
+	for i := range entries {
+		for j := range entries[i].Requires {
+			entries[i].Requires[j].ConvertVersionToMetadata()
+		}
 	}
+	return buildpack.Group{Group: bps}, platform.BuildPlan{Entries: entries}, err
 }
 
-func (r *Require) convertVersionToMetadata() {
-	if r.Version != "" {
-		if r.Metadata == nil {
-			r.Metadata = make(map[string]interface{})
+func (d *Detector) detectOrder(order buildpack.Order, done, next []buildpack.GroupBuildpack, optional bool, wg *sync.WaitGroup) ([]buildpack.GroupBuildpack, []platform.BuildPlanEntry, error) {
+	ngroup := buildpack.Group{Group: next}
+	buildpackErr := false
+	for _, group := range order {
+		// FIXME: double-check slice safety here
+		found, plan, err := d.detectGroup(group.Append(ngroup), done, wg)
+		if err == ErrBuildpack {
+			buildpackErr = true
+		}
+		if err == ErrFailedDetection || err == ErrBuildpack {
+			wg = &sync.WaitGroup{}
+			continue
 		}
-		r.Metadata["version"] = r.Version
-		r.Version = ""
+		return found, plan, err
+	}
+	if optional {
+		return d.detectGroup(ngroup, done, wg)
 	}
-}
 
-func (r *Require) hasInconsistentVersions() bool {
-	if version, ok := r.Metadata["version"]; ok {
-		return r.Version != "" && r.Version != version
+	if buildpackErr {
+		return nil, nil, ErrBuildpack
 	}
-	return false
+	return nil, nil, ErrFailedDetection
 }
 
-func (r *Require) hasDoublySpecifiedVersions() bool {
-	if _, ok := r.Metadata["version"]; ok {
-		return r.Version != ""
+func (d *Detector) detectGroup(group buildpack.Group, done []buildpack.GroupBuildpack, wg *sync.WaitGroup) ([]buildpack.GroupBuildpack, []platform.BuildPlanEntry, error) {
+	for i, groupBp := range group.Group {
+		key := groupBp.String()
+		if hasID(done, groupBp.ID) {
+			continue
+		}
+
+		bp, err := d.Store.Lookup(groupBp.ID, groupBp.Version)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		bpDesc := bp.ConfigFile()
+		groupBp.API = bpDesc.API
+		groupBp.Homepage = bpDesc.Buildpack.Homepage
+
+		if bpDesc.IsMetaBuildpack() {
+			// TODO: double-check slice safety here
+			// FIXME: cyclical references lead to infinite recursion
+			return d.detectOrder(bpDesc.Order, done, group.Group[i+1:], groupBp.Optional, wg)
+		}
+		done = append(done, groupBp)
+		wg.Add(1)
+		go func(key string, bp Buildpack) {
+			if _, ok := d.Runs.Load(key); !ok {
+				d.Runs.Store(key, bp.Detect(&d.DetectConfig))
+			}
+			wg.Done()
+		}(key, bp)
 	}
-	return false
-}
 
-func (r *Require) hasTopLevelVersions() bool {
-	return r.Version != ""
+	wg.Wait()
+
+	return d.Resolver.Resolve(done, d.Runs)
 }
 
-type Provide struct {
-	Name string `toml:"name"`
+func hasID(bps []buildpack.GroupBuildpack, id string) bool {
+	for _, bp := range bps {
+		if bp.ID == id {
+			return true
+		}
+	}
+	return false
 }
 
-type DetectConfig struct {
-	FullEnv       []string
-	ClearEnv      []string
-	AppDir        string
-	PlatformDir   string
-	BuildpacksDir string
-	Logger        Logger
-	runs          *sync.Map
+type DefaultResolver struct {
```

Comment:
There is a `MockResolver` and `Resolver` is the interface. I've seen `DefaultX` as a pattern elsewhere, but I'm open to choosing a different name if you think it would be clearer.

Reply:
Thank you for your clarification. I missed the full picture.
I think it's ok to keep it as `DefaultResolver`.